spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java

```diff
@@ -1159,16 +1192,33 @@ public Message doInRabbit(Channel channel) throws Exception {
 				if (logger.isDebugEnabled()) {
 					logger.debug("Sending message with tag " + messageTag);
 				}
-				doSend(channel, exchange, routingKey, message, correlationData);
-				LinkedBlockingQueue<Message> replyHandoff = pendingReply.getQueue();
-				Message reply = (replyTimeout < 0) ? replyHandoff.take() : replyHandoff.poll(replyTimeout,
-						TimeUnit.MILLISECONDS);
-				RabbitTemplate.this.replyHolder.remove(messageTag);
+				Message reply = null;
+				try {
+					reply = exchangeMessages(exchange, routingKey, message, correlationData, channel, pendingReply,
+							messageTag);
+				}
+				finally {
+					RabbitTemplate.this.replyHolder.remove(messageTag);
+				}
 				return reply;
 			}
 		}, obtainTargetConnectionFactoryIfNecessary(this.sendConnectionFactorySelectorExpression, message));
 	}
 
+	private Message exchangeMessages(final String exchange, final String routingKey, final Message message,
+			final CorrelationData correlationData, Channel channel, final PendingReply pendingReply, String messageTag)
+			throws Exception, InterruptedException {
+		Message reply;
```

Comment:
Hopefully, to clarify...

With direct-replyTo (no listener), the thread handoff from the reply to the request was done via a BlockingQueue within the method scope. With a listener container, we had a map of pending replies, keyed by the correlation id because the reply does not come within the method scope.

Now, both use the map of pending replies so that if a return is received, an exception can be queued to the requestor.

Hope that helps.